README.md
<p align="center">
  <img width="25%" height="25%" src="img/pontstrader_logo.jpg">
</p>
pontstrader is a trading script running on python2.7 which adds some extra features to the default Bittrex trading possibilities trough the Bittrex webinterface.

Due to it being written in Python, it can run on any device such as: Windows, Mac, Linux, Raspberry Pi, Phone and NAS.

!!! PONTSTRADER IS NOT AN AUTOMATED TRADING BOT !!!

!!! This script is here to help you with your trades, we are not responsible for losses !!!

!!! SCRIPT IS NOT USABLE IF YOU ARE NOT A PAYING CUSTOMER !!!

Index:
  - [Features](#features)
  - [Installation](docs/Installation.md)
  - [Upgrade version or re-configure](docs/Upgradereconfigure.md)
  - [Releasenotes](docs/Releasenotes.md)

Join us on Telegram for more information or if you would like to purchase pontstrader for just 10 dollar:
https://t.me/pontstrader

## What can users do with pontstrader

1. Buy cryptocurrencies (Bittrex)
2. Sell cryptocurrencies (Bittrex)
3. Buy cryptocurrencies and immidiatly add a sell order with a multiplier (Bittrex)
4. Check their balances (Bittrex)
5. Check their orderbook (Bittrex)
6. Watch a cryptocurrencies in real-time 0.5 seconds (Bittrex)
7. Withdraw cryptocurrencies (Bittrex)
8. Check arbitrage oppertunaties between exchanges (Bittrex, HitBTC, Binance, Bitfinex, Poloniex)
9. Use Trailing Stop Loss, which is normally not possible on most exchanges (Bittrex)

Some of the above functions require the script to run on a system which is on 24/7, this is due to the fact the script is actively (every 0.5 seconds) quering the pontstrader redis database for price changes.
Currently the following functions require a 24/7 running system:

- Trailing Stop Loss

## What can users expect from pontstrader in the near future

1. Stop Loss + Take Profit function which allows you to set not just one but multiple tresholds, which means the script will not only sell when it hits your stop loss treshold but will also sell when it hits your profit tresholds. Normally you can only choose one on Bittrex. (this will also require a 24/7 running system)
2. Stop Loss + Take Profit multiple sell tresholds. Same as the above, but then a 50/50 split in selling tresholds. For example you want to sell 50% at 5% profit, and the other 50% at 10%. (this will also require a 24/7 running system)
3. Multiple exchange support on every function, currently only Bittrex is supported on most of them
4. And much more!

Join us on Telegram for more information or if you would like to purchase pontstrader for just 10 dollar:
https://t.me/pontstrader

## Features

#### 1. Buy
The buy function allows you to buy any coin on any market supported on the Bittrex exchange.
  - Choose a market (BTC, ETH or USDT)
  - Choose a currency such as NEO, LTC or LSK
  - Price information will be shown, also shows you how much you can buy at current Last, Ask and Bid rates
  - You are asked to choose to buy for Last, Ask, Bid or Custom
  - Choose the amount you would like to buy
  - Recheck your input and confirm the buy if everything looks find
<img width="25%" height="25%" src="img/buy.png">

  

#### 2. Sell
The sell function allows you to sell any coin on any market supported on the Bittrex exchange.
  - Upon starting this function your balances will be checked.
  - Choose the currency by providing its corresponding number.
  - Choose the market where you would like to sell the coin on (Make sure the market exists).
  - Price information will be shown
  - You are asked to choose to sell for Last, Ask, Bid or Custom
  - Choose the amount you would like to sell
  - Recheck your input and confirm the sell if everything looks find
<img width="25%" height="25%" src="img/sell.png">

#### 3. Buy and Sell
The buy and sell function allows you to buy any coin on any market supported on the Bittrex exchange, after buying create a sell order with a multiplier
  - Choose a market (BTC, ETH or USDT)
  - Choose a currency such as NEO, LTC or LSK
  - Price information will be shown, also shows you how much you can buy at current Last, Ask and Bid rates
  - You are asked to choose to buy for Last, Ask, Bid or Custom
  - Choose the amount you would like to buy
  - Choose the multiplier (1.1 = 10%)
  - Recheck your input and confirm the buy if everything looks find
  - The script will buy your currency, and create a sell order after the buy is completed
<img width="25%" height="25%" src="img/buysell.png">

#### 4. Balances
The balances function is pretty straight forward, it shows all your balances upon starting it as you can see on this screenshot below.

<img width="25%" height="25%" src="img/balances.png">

#### 5. Orderbook
The orderbook function is pretty straight forward, it shows all your open orders upon starting it as you can see on this screenshot below.

<img width="25%" height="25%" src="img/orderbook.png">

#### 6. Watch
The watch function allows you to watch a coin in real-time, quering the pontstrader redis database server every 0.5 for price updates. Only price changes are shown, check the screenshot.

<img width="25%" height="25%" src="img/watch.png">

#### 7. Withdraw
The withdraw function allows you to withdraw coins towards other wallets, also pretty straight forward and easy to use.
Dont forget to add a Payment ID if required.

<img width="25%" height="25%" src="img/withdraw.png">

#### 8. Arbitrage
The arbitrage function allows you to show price differences between 5 different exchanges, currently supported exchanges are:
Bittrex, HitBTC, Binance, Bitfinex and Poloniex
  - You can choose to show oppertunaties only from a specific market
  - You can choose to show oppertunaties only from a specific exchange

<img width="25%" height="25%" src="img/arbitrage.png">

#### 9. Trailing Stop Loss (Requires 24/7 running device)
The trailing stop loss function allows you to trade currencies on a more advanced matter, if you do not know what trailing stop loss is please have a look at: https://www.investopedia.com/articles/trading/08/trailing-stop-loss.asp

Upon selling a coin, either with loss or profit... a push notification will be send towards your phone if you've enabled pushover or pushbullet (which is awesome btw!)

Please keep in mind this function requires the system running this to be on 24/7, because this script will look for price differences every 0.5 seconds and is not able to do this when its not on. A recommendation will be a power efficient raspberry pi with for example Raspbian, you can run pontstrader is a screen or tmux session.

Small example:
If you set a trailing stop loss treshold to 10% your trade will immidiatly be sold when it falls 10% straight away, but if it climbs to 5% profit and afterwards falls 10% it will stop at -5% instead of minus 10%, this results in less loss.
If the price starts climbing and it goes towards 20%, the trailing stop loss will be set to 10%. When it starts falling it will sell on 10% which results in less risk and restless sleep for you :)

<img width="25%" height="25%" src="img/trailing1.png">
<img width="25%" height="25%" src="img/trailing2.png">

#### 10. Take Profit (Requires 24/7 running device)
The take profit function is actually the same as the Sell option, only the difference is that this script will keep track of the sell and notify you by a push message when the sell have happend.

Because the script is keeping track of the sell, you will need to run this on a device which is on 24/7.

<img width="25%" height="25%" src="img/takeprofit1.png">
<img width="25%" height="25%" src="img/takeprofit2.png">

#### 11. Stop Loss + Take Profit (Requires 24/7 running device)
The take profit function is actually the same as the Sell option, including a stop loss function which allows you to set not only one but both of them.

Same as the other 24/7 functions this function will notify you on changes through a push message.

<img width="25%" height="25%" src="img/stoplosstakeprofit1.png">
<img width="25%" height="25%" src="img/stoplosstakeprofit2.png">

#### 12. Trailing + Take Profit (Requires 24/7 running device)
This function is the same as the trailing stop loss function, the only difference is... is that this function will never ever sell your coin when its down, it will only trigger a sell when the trailing stop loss threshold is in profit.

<img width="25%" height="25%" src="img/trailingtakeprofit1.png">
<img width="25%" height="25%" src="img/trailingtakeprofit2.png">

Join us on Telegram for more information or if you would like to purchase pontstrader for just 10 dollar:
https://t.me/pontstrader


Releasenotes.md
# v3.1
- Arbitrage hot fix (float error)

v3
- Removed the automated watch function, now allows quiting during watch again
- All public API data is now retrieved from the pontstrader servers taken from a Redis cluster, allowing more speed, less limitations and flexibility for extra functions in the future
- Allot of things done in the background such as:
* Monitoring (collect, visualize, notify)
* Infrastructure (server migration make it ready for Redis)
* Backup (snapshots and backups)
* Security (firewalls)
* Allot other small stuff

v2.2
- Added refresh option after checking trailing stop loss status
- Revamped the progress output on the trailing stop function, its now possible the check all status updates
v2.1
- Retrieving public market data for any function now runs in the background which improves speed even more
- Fixed error output for Pushover and Pushbullet
- Removed usage of CTRL-C in the script

v2.0 (With trailing stop trading BETA)
- New trading function: Trailing Stop for Bittrex (check the docs for more information)
- New feature: Push notifications on your phone with Pushover or Pushbullet (Trailing stop function)
- Arbitrage function out of BETA - Arbitrage speed improvements, data retrieving is now running in the background
- Removed exit's (BETA) on the arbitrage function, script does not exit anymore
- Fixed a bug where closing the function would give a API error
- Fixed a bug which made the script crash if you had no orders in the orderbook
- Fixed a bug where the arbitrage function would crash if you use the menu to quick
- Fixed a error caused by choosing 0 in the sell, orderbook and withdraw function
- Watch function is not killed on API error anymore during watch
- Arbitrage function is not killed on API error anymore during watch
- Added function to buy/sell and buysell that will allow a watch after the order to see if its filled
- Added missing BTC value for the BTC currency in the balance function

v1.20 (With arbitrage BETA)
- Version check (e.g. run.py version)
- Cancelling a watch allows you to quickly watch a new coin
- Removed all sys.exit's which means the script only dies in API errors or invalid input now
- Added back to Main Menu options
- Removal of unrealistic arbitrage oppertunaties (1000% +)
- Balance gets automatically refreshed every 30 seconds
- Fixed arbitrage polo zero float error (Thanks Robbert)

v1.14 (With arbitrage BETA)
- Added option to choose market with arbitrage
- Added option to choose from how many percentage you want entries to show

v1.13 (With arbitrage BETA)
- Added bitfinex to the arbitrage function

v1.12
- Withdraw function release

v1.11 BETA (With arbitrage and withdraw BETA)
- Arbitrage results now have numbers for future usage (automation)
- Withdraw function now allows the use of payment ID's

v1.1 BETA
- Arbitrage BETA
- Withdraw BETA (does not work with Payment ID's yet)
- Buy / Watch / Sell: Fixed error when using lowercase characters as currency
- Arbitrage: Added support for Binance
- Colors: Removed the ansicon requirement for windows (python-colorama)
- Added an install wizard when the script is used for the first time

v1.0
- Public release

v0.9
- Balance: Balance now shows total estimated value in BTC and Dollar
- Balance: You can now choose out of the 3 possible Bittrex markets instead of typing them

v0.8
- Balance / Sell: Added more information during a buy or sell, price/fee/minimum etc.

- Orderbook: Open orders menu option to allow overview and removal of open orders
v0.7
- Buy / Sell functions revamped
- Buy and Sell with custom values instead of only Last price
- Sell directly from your wallet, instead of manually inserting currency to sell (after all, you can only sell what you own right?)

v0.6
- NEW: Balance function

v0.5
- NEW: Orderbook

v0.4
- NEW: Watch function

v0.3
- NEW: Buy + Sell funtion

v0.2
- NEW: Sell function

v0.1
- NEW: Buy function


functions/menu.py
#!/usr/bin/env python

def menu(apikey, apisecret, pushover_user, pushover_app, pushbullet_token, redis_password):

  import sys
  import time
  import threading
  import os
  from buy import buy
  from sell import sell
  from buysell import buysell
  from balances import balances
  from orderbook import orderbook
  from watch import watch
  from withdraw import withdraw
  from arbitrage import arbitrage
  from trailing import trailing
  from takeprofit import takeprofit
  from stoplosstakeprofit import stoplosstakeprofit
  from trailingtakeprofit import trailingtakeprofit
  from colorama import Fore, Back, Style, init
  init(autoreset=True)

  while True:
    print (30 * '-')
    print(Fore.GREEN + 'P O N T S T R A D E R . C O M')
    print (30 * '-')
    print(Fore.YELLOW + '1. Buy')
    print(Fore.YELLOW + '2. Sell')
    print(Fore.YELLOW + '3. Buy and Sell')
    print(Fore.YELLOW + '4. Balances')
    print(Fore.YELLOW + '5. Orderbook')
    print(Fore.YELLOW + '6. Watch coin')
    print(Fore.YELLOW + '7. Withdraw')
    print(Fore.YELLOW + '8. Arbitrage')
    print(Fore.YELLOW + '9. Trailing stop (24/7)')
    print(Fore.YELLOW + '10. Take Profit (BETA + 24/7)')
    print(Fore.YELLOW + '11. Stop Loss + Take Profit (BETA + 24/7)')
    print(Fore.YELLOW + '12. Trailing + Take Profit (BETA + 24/7)')
    print(Fore.RED +'13. Exit')
    print (30 * '-')

    try:
      choice = raw_input('Enter your choice [1-13] : ')
      choice = int(choice)
    except:
      print 'Invalid number. Try again...'

    # BUY
    if choice == 1:
      buy(apikey, apisecret, redis_password)

    # SELL
    elif choice == 2:
      sell(apikey, apisecret, redis_password)

    # BUY AND SELL
    elif choice == 3:
      buysell(apikey, apisecret, redis_password)

    # SHOW WALLETS
    elif choice == 4:
      balances(apikey, apisecret, redis_password)

    # OPEN ORDERS
    elif choice == 5:
      orderbook(apikey, apisecret, redis_password)

    # WATCH
    elif choice == 6:
      watch(apikey, apisecret, redis_password)

    # WITHDRAW
    elif choice == 7:
      withdraw(apikey, apisecret)

    # ARBITRAGE
    elif choice == 8:
      arbitrage(redis_password)

    # TRAILING
    elif choice == 9:
      trailing(apikey, apisecret, pushover_user, pushover_app, pushbullet_token, redis_password)

    # TAKE PROFIT
    elif choice == 10:
      takeprofit(apikey, apisecret, pushover_user, pushover_app, pushbullet_token, redis_password)

    # STOP LOSS TAKE PROFIT
    elif choice == 11:
      stoplosstakeprofit(apikey, apisecret, pushover_user, pushover_app, pushbullet_token, redis_password)

    # TRAILING TAKE PROFIT
    elif choice == 12:
      trailingtakeprofit(apikey, apisecret, pushover_user, pushover_app, pushbullet_token, redis_password)

    # EXIT
    elif choice == 13:
      count = threading.activeCount()
      if count > 1:
        threads = threading.enumerate()
        thread_counter = 0
        for t in threading.enumerate():
          if 'arbitrage' in t.name:
            pass
          elif 'Main' in t.name:
            pass
          else:
            thread_counter += 1
        if thread_counter > 0:
          print (Fore.YELLOW +'WARNING: There are currently {0} active trade(s), are you sure you want to exit?'.format(thread_counter))
          print(Fore.GREEN +'1. yes')
          print(Fore.RED +'2. no')
          try:
            yes_no = raw_input('Enter your choice [1-2] : ')
            yes_no = int(yes_no)
          except:
            print 'Invalid number... going back to Main Menu'
          if yes_no == 1:
            print 'Exiting...'
            sys.exit()
          elif yes_no == 2:
            print 'Good... going back to Main Menu'
          else:
            print 'Invalid number... going back to Main Menu'
        else:
          print 'Exiting...'
          sys.exit()
      else:
        print 'Exiting...'
        sys.exit()

    # ELSE EXIT
    else:
      print 'Invalid number. Try again...'
